### PR TITLE
[WIP] tree from parents

### PR DESCRIPTION
**Pull request by pradkrish**

### CODE REVIEW

**Comment by GroteGnoom** [comment]
For easier reference:
Fixes #1840

**Comment by GroteGnoom** [comment]
I don't think I have the time to go through everything thoroughly,  but I'll at least add some comments :)

File: src/constructors/regular.c [hunk 1]
```diff
@@ -420,6 +420,81 @@ igraph_error_t igraph_tree(igraph_t *graph, igraph_integer_t n, igraph_integer_t
...
+igraph_error_t igraph_tree_from_parent_vector(igraph_t *graph, igraph_vector_int_t* parents,
+                igraph_tree_mode_t type) {
+
+    igraph_vector_int_t edges = IGRAPH_VECTOR_NULL;
+    igraph_vector_int_t visited = IGRAPH_VECTOR_NULL;
+    igraph_integer_t idx = 0;
+    igraph_integer_t to = 1;
+
+    if (igraph_vector_int_empty(parents)) {
+        IGRAPH_ERROR("Parent vector is empty", IGRAPH_EINVAL);
```
Comment (GroteGnoom) [inline src/constructors/regular.c:432]: Please end error sentences with a .

File: tests/unit/igraph_tree_from_parent_vector.c [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+*/
+
+#include <igraph.h>
+
+#include "test_utilities.inc"
+
+int main() {
+  igraph_t graph;
+  igraph_vector_int_t parents;
+  igraph_vector_int_init(&parents, 5);
```
Comment (GroteGnoom) [inline tests/unit/igraph_tree_from_parent_vector.c:26]: you can use vector_int_init_int to immediately put values into the vector.

File: src/constructors/regular.c [hunk 3]
```diff
@@ -420,6 +420,81 @@ igraph_error_t igraph_tree(igraph_t *graph, igraph_integer_t n, igraph_integer_t
...
+      child_node = v;
+
+      if (!VECTOR(visited)[child_node]) {
+        VECTOR(visited)[child_node] = 1;
+        continue;
+      }
+
+      parent_node = VECTOR(*parents)[v];
+      if (parent_node>= n) {
+        IGRAPH_ERROR("Invalid entry in parents", IGRAPH_EINVAL);
```
Comment (GroteGnoom) [inline src/constructors/regular.c:454]: Please use ERRORF to indicate what you found and what you expected, this can make it easier to debug problems.

**Comment by GroteGnoom** [comment]
You don't seem to use the visited vector to check for cycles. Maybe you can add a test case for for a 2-edge cycle, which I think is detected correctly, and then a 3-edge cycle, which I think is not detected properly.

File: src/constructors/regular.c [hunk 4]
```diff
@@ -420,6 +420,81 @@ igraph_error_t igraph_tree(igraph_t *graph, igraph_integer_t n, igraph_integer_t
...
+    }
+
+    igraph_integer_t n = igraph_vector_int_size(parents);
+    IGRAPH_VECTOR_INT_INIT_FINALLY(&visited, n);
+    IGRAPH_VECTOR_INT_INIT_FINALLY(&edges, 2 * (n - 1));
+
+    // check if parents contains any cycles
+    igraph_integer_t child_node, parent_node, v;
+    for (int v=n-1; v>=0; --v) {
+      child_node = v;
```
Comment (GroteGnoom) [inline src/constructors/regular.c:445]: You sometimes have an indentation of 2 instead of 4. This makes it hard to see blocks.

File: src/constructors/regular.c [hunk 5]
```diff
@@ -420,6 +420,81 @@ igraph_error_t igraph_tree(igraph_t *graph, igraph_integer_t n, igraph_integer_t
     return IGRAPH_SUCCESS;
 }
 
+igraph_error_t igraph_tree_from_parent_vector(igraph_t *graph, igraph_vector_int_t* parents,
+                igraph_tree_mode_t type) {
+
+    igraph_vector_int_t edges = IGRAPH_VECTOR_NULL;
+    igraph_vector_int_t visited = IGRAPH_VECTOR_NULL;
```
Comment (szhorvat) [inline src/constructors/regular.c:427]: I suggest not to initialize to values you will never use. This tends to interfere with static analysis tools.

File: src/constructors/regular.c [hunk 6]
```diff
@@ -420,6 +420,81 @@ igraph_error_t igraph_tree(igraph_t *graph, igraph_integer_t n, igraph_integer_t
...
+    igraph_integer_t idx = 0;
+    igraph_integer_t to = 1;
+
+    if (igraph_vector_int_empty(parents)) {
```
Comment (szhorvat) [inline src/constructors/regular.c:431]: The number of vertices in the graph will be the same as the length of the `parents` vector. Thus, an empty parents vector can represent the null graph, which is a forest, thus a valid result.

File: src/constructors/regular.c [hunk 7]
```diff
@@ -420,6 +420,81 @@ igraph_error_t igraph_tree(igraph_t *graph, igraph_integer_t n, igraph_integer_t
...
+        IGRAPH_ERROR("Invalid mode argument", IGRAPH_EINVMODE);
+    }
+
+    igraph_integer_t n = igraph_vector_int_size(parents);
+    IGRAPH_VECTOR_INT_INIT_FINALLY(&visited, n);
+    IGRAPH_VECTOR_INT_INIT_FINALLY(&edges, 2 * (n - 1));
+
+    // check if parents contains any cycles
+    igraph_integer_t child_node, parent_node, v;
+    for (int v=n-1; v>=0; --v) {
```
Comment (szhorvat) [inline src/constructors/regular.c:444]: We still do not make use of all features in new C standards, in order to maintain compatibility with some compilers. Thus, for now, please do not use declarations directly in the for loop. (We may start accepting this later, but it has not happened yet.)

Always use `igraph_integer_t` for vertex IDs, not `int`.

**Comment by szhorvat** [comment]
Happy new year! Let us know if you are still working on this project.

**Comment by szhorvat** [comment]
@pradkrish Are you still planning to work on this?

**Comment by pradkrish** [comment]
@szhorvat Yes although I haven't even had time to even log into my github account for some weeks now. Omicron, school closures etc. Thanks for the nudge, I will try to get back to this as soon as I can.

**Comment by ntamas** [comment]
I have merged the recent changes of the `develop` branch to this PR. @pradkrish if you resume working on this PR, you should run `git pull` first in your own copy to get the changes.

**Comment by pradkrish** [comment]
Hello, I am back now to resume working on this. I need to refresh my mind where I left off. It looks like there is some discussion to follow up. Should I still continue with this or are there other plans?

**Comment by ntamas** [comment]
Resolved conflicts with the current `develop` branch.

@szhorvat is this for 0.10.0 or 0.10.x (where x > 0)? It does not seem like breaking the API anywhere so this could be targeted at 0.10.x in my opinion.

**Comment by szhorvat** [comment]
Retargeted. I want to have this in 0.10.x and I plan to do it myself.

**Comment by ntamas** [comment]
Removed the stale label as @szhorvat has plans to do this. Also resolved conflicts with `develop`. Maybe it should be retargeted to `master` now that we mostly work there again?

**Comment by szhorvat** [comment]
Continued in #2266
